uniffi/src/support/mod.rs

```diff
@@ -194,37 +195,33 @@ unsafe impl<T: Liftable + Lowerable> ViaFfi for Vec<T> {
     }
 }
 
-unsafe impl<'a> ViaFfi for &'a str {
-    type Value = ffi_support::FfiStr<'a>;
-    fn try_from_ffi_value(v: Self::Value) -> Result<Self> {
-        Ok(v.as_str())
-    }
-
-    // This should never happen since there is asymmetry with strings
-    // We typically go FfiStr -> &str for argumetns and
-    // String -> *mut c_char for return values.
-    // We panic for now, the ViaFfi triat will be
-    // broken down to IntoFfi and TryFromFfi to prevent us from having
-    // to add this
-    fn into_ffi_value(&self) -> Self::Value {
-        panic!("Invalid conversion. into_ffi_value should not be called on a &str, a String -> *mut c_char conversion should be used instead.")
-    }
-}
-
+/// Support for passing Strings back and forth across the FFI.
+///
+/// Unlike many other implementations of `ViaFfi`, this passes a pointer rather
+/// than copying the data from one side to the other. This is a safety hazard,
+/// but turns out to be pretty nice for useability.
+///
+/// (In practice, we do end up copying the data, the copying just happens on
+/// the foreign language side rather than here in the rust code.)
 unsafe impl ViaFfi for String {
     type Value = *mut std::os::raw::c_char;
-    fn into_ffi_value(&self) -> Self::Value {
+
+    // This returns a raw pointer to the underlying bytes, so it's very important
+    // that it consume ownership of the String, which is relinquished to the foreign
+    // language code (and can be returned by it passing the pointer back).
```

Comment:
Also this meant that I had to revert `into_ffi_value` to consuming `self` by value.